### ISSUE TITLE
Folder as Workspace causes very slow startup

### ISSUE THREAD

**Issue by byronigoe** [notepad-plus-plus/notepad-plus-plus#1566]
Thank you for the Folder as Workspace issue, but there is a performance problem with large metadata.

It appears that Notepad++ is doing a deep scan of folders and files during startup when you use "Folder as Workspace". I have 2 folders, each contains about 8000 folders (average depth of ~5) and 30000 files. This causes NPP to take over a minute to be usable after initial launch.

Is this startup scan necessary? If so, why? Can we have an option to turn it off?

**Comment by Ghasan** [comment]
Did you try to close the FaW before closing Notepad++, and reopening it again?

**Comment by byronigoe** [comment]
If I close FaW, close Notepad++, open Notepad++, then it opens quickly. Then if I select View->Folder as Workspace, it takes over a minute.

**Comment by sanastasiou** [comment]
IO stuff should be done asynchronously. Seems not to be the case here.

**Comment by byronigoe** [comment]
Am I the only one experiencing this? I do development of web apps, and having Bower dependencies creates LOTS of subdirectories. If FaW can't be easily fixed, is there a workaround?

**Comment by byronigoe** [comment]
Now I have 7 folders at the top level, and startup takes 2.5 minutes. What processing is being done? It should be done lazily, or in the background.

**Comment by wejoey** [comment]
Simply to mention that I experience the same issue with the Folder as Workspace. In my case, it takes about 5 to 7 seconds to open for about 12k files in 6k folders.

**Comment by sanastasiou** [comment]
That's not an issue at all if you never close Npp. A far more critical and annoying issue is, that with many changes in the filesystem Npp just crashes and / or displays "Queue Overflow" popups.. That is annoying and unreliable.

**Comment by byronigoe** [comment]
@sanastasiou Are you joking by saying "never close NPP"? Even if I restart my computer only once a week, waiting 150 seconds for the application to be usable is not acceptable. If Photoshop starts up faster than a text editor, there's a problem. The issue you mention sounds like it should be submitted separately, instead of diverting this thread. Can anyone please answer my previous question:

> What processing is being done? It should be done lazily, or in the background.

**Comment by sanastasiou** [comment]
@byronigoe I am not joking.. I am just using hibernation to prevent it from ever closing.. Until this is fixed, and it will be a while ( months ) if it ever gets fixed this is a simple workaround.

I am more worried about the constant crashes when u update a lot of files together.. as when you compile/build something and the annoying "queue overflow" message box that comes with it.. It's really unstable.

**Comment by byronigoe** [comment]
I've switched from NPP to Atom, which starts up in 10 seconds for the same folder structure.

**Comment by SonnyAD** [comment]
I don't want to switch Atom.. If you could find a way to fix that, that would be awesome :)

**Comment by prionkor** [comment]
Any news on this? I am using notepad++ with wine in Ubuntu. I have to wait 15 min everytime I start notepad :/

**Comment by Daimanu06** [comment]
Same problem here, as soon as the "folder as workspace" menu is up, N++ is very slow, even without network folder. No speed issue if closed. (N++ v. 7.5.3 x64)

**Comment by joelBrun** [comment]
The issue is still present in notepad++ 7.5.1 (32-bit)

**Comment by donho** [comment]
Close Folder as Workspace panel will remedy this problem.

**Comment by cklee75** [comment]
The issue is still present in notepad++ 7.8.5 (32-bit). If it can't be done lazily, or in the background, at least let users have option to disable FaW during startup.

**Comment by patiqs** [comment]
Same issue here. Why is this ticket closed?

**Comment by MichaelFerrier** [comment]
Just ran into this issue and I can see by this ticket it won't be fixed anytime soon. Switching to Atom.

**Comment by PsykomantaBrain** [comment]
Adding my +1 here. It would be great to be able to use this feature, but the slow startup makes it kind of unusable.

**Comment by ArkadiuszMichalski** [comment]
Report also in forum:
https://community.notepad-plus-plus.org/topic/10669/why-does-the-app-take-so-long-to-start/11

**Comment by donho** [comment]
It should be improved by https://github.com/notepad-plus-plus/notepad-plus-plus/commit/ca07ac69c7816804601b8420c1f1c22dd7fc77aa. Any confirmation?

**Comment by donho** [comment]
If there's no one care about testing it, this request can be closed.

**Comment by MichaelFerrier** [comment]
Personally, I don't test it because the issue was bad enough to cause me to use a different text editor. I do keep an eye on this thread in hopes it will be fixed so I can give Notepad++ another try, though.

**Comment by mm-nv** [comment]
I'm like @MichaelFerrir.  I'll give it a try to see if it actually works like all the other editors out there.  I mean, it was so slow it was useless.  I use n++ everyday for everything except coding projects because of the poor performance in the explorer.  I hope it works now.

**Comment by mere-human** [comment]
@mm-nv  @MichaelFerrier Have you tried [Project Panels](https://npp-user-manual.org/docs/session/#project-panels)?

**Comment by phpjunkie420** [comment]
I had this issue and was able to fix it. I have a program that is generating allot of files as it scans through the RAM and the CPU Registers. This folder ended up with 9000+ files in it. After deleting this folder notepad++ started working again.

Anyway, this is most likely the issue. Notepad++ is probably processing far to many files for Folder as Workspace to load the tree view. I was also reading about [Project Panel](https://npp-user-manual.org/docs/session/#project-panels)s that can be found in View->Project Panels. You can use Project Panels to create a folder tree view independent of the Windows filesystem hierarchy.

**Comment by irmlab** [comment]
(wobotii) This problem also seems to occur due to circular symbolic links.

Exactly ! I think this is the correct diagnostic. Even if symbolic links are not circular, 
In my case, I use multiple symlinks that point to a central backup directory with hundred of files.
It appear that each symlink directory is parsed again and again. 
As a result, NPP parse 300 files multiplied by 15 symlinks directories and NPP lasts about 4 minutes (!)

The solution would be easy though.
Simply add some filter using a folder mask (i.e.: "bak*/") so the matching folders will be delayed on startup.
Delayed means : collapsed until the user click on them.

**Comment by darthglowball** [comment]
My issue is worse: Notepad++ (v8.9.1, 64-bit, Windows 10) freezes indefinitely when there are symbolic links in at least the top-level folders! Some of the symbolic links have an invalid target (by design; I use some of them as key value pairs that display as such in a reparse-point-aware file manager). There might be other conditions that create the same problem, I have not tried it. I can try and tell exact conditions if asked. (I know that my Notepad++ version is a bit outdated; but has the problem been solved since then?). As I heavily use reparse points, and am going to for the rest of my life, and will build file system object abstractions on top of them, it is very disappointing that the "Folder as Workspace" is unusable! Will this be looked into? (if anyone will, please, after it, do an effort to make nice icons for reparse points, and have the target's path display next to it without having to mouse over it (if this is not the case already; can not test it because it freezes Notepad++ as I mentioned) (perhaps add a toggle button to hide/show the target path), and when clicking the node, it selects the target and pans the camera to it (instead of expanding a tree copy of the target); and, have relative paths display as relative (I mean duh: do not do what Windows 10 Explorer does, whereby it converts them to absolute paths!); thank you in advance!!!)

**Comment by pryrt** [comment]
@darthglowball ,
> freezes indefinitely when there are symbolic links in at least the top-level folders! Some of the symbolic links have an invalid target

Is your problem with the symbolic links _only_ when there is one or more link that points to an invalid target?  Or does it also happen when the only symbolic link has a valid target?  If it's the latter, then that is definitely a bug.  If it's just with an invalid target (or the circular target from earlier in the discussion), then ... well, since it completely freezes, it's still a bug, though I hate suggesting that a "bug" needs to be fixed in N++ when someone is abusing the filesystem in ways it wasn't designed to be used.

> Will this be looked into?

The issue was originally closed because no one cared to report whether the fix that the developer tried helped things.  Since you seem interrested, I will re-open to issue, and ping to see if @donho is willing to try to add back in his original attempt at a fix into the modern codebase, for you to test.  However, he's already closed it once because people stopped caring... so I would advise you to watch your GH notifications (maybe even enable the automatic email in your notification settings, if you don't visit GH often), and reply to questions -- otherwise, he's likely to close it again, and ignore all further attempts to revive it.

> if anyone will, please, after it, do an effort to make nice icons for reparse points, and have the target's path display next to it without having to mouse over it  ...

You should create a separate feature request for that, as that's not related to the slow startup / freeze that's at issue here.

**Comment by darthglowball** [comment]
@pryrt I am replying late because I wrote a long response to you about the justification for "link loops" and "invalid link targets", but I realized at the last minute that "link loops" is ambiguous.

There are at least two interpretations:
1) _single user action infinite loop_: in a link chain, the last link's target pointing to the first link's path, creating an infinite OS shell loop by entering the first link once (well, if the OS is unable to detect this)
2) _multiple user actions finite loop_: link target equal to parent folder (`.`) or some amount of parent up's (`..\`, `..\..\`, etc), creating a finite OS shell loop by entering the first link a fixed amount of times depending on repeated user actions, possibly involving entering multiple folders for one loop iteration to get back to the first link (eg. `a/a/a/a/a/a`, `a/b/c/a/b/c`)

Which ones do you like?

**Comment by pryrt** [comment]
> Which ones do you like

Since there aren't "standard" definitions, I don't think it matters, as long as you are consistent and say which you mean:

What I guess would be important, for the developer (not for me: I have no clue where this problem would be, or how to fix it; I am just offering advice for how to successfully get your point across)

- be very explicit: 
    - does the problem occur when the only link file is a "normal" link?  
    - or does it require just any bad link (like a link to a destination that no longer exists)? 
    - or does it only happen when there is a link that causes a circle (where you can define that "circle" or "link loop" however you want, as long as you are clear)

- be willing to try out potential code fixes if the developer goes to the effort of redoing the fix that no one bothered to try five years ago

**Comment by donho** [comment]
@pryrt @darthglowball 
Please create a new issue with the step by step instructions - once I can reproduce the bug, I'll try to fix it.

**Comment by darthglowball** [comment]
@donho I am sorry, I have been mistaken with the folder tree I analyzed. The freeze does not occur there with the root folder containing one symbolic link with a valid or invalid target (correct and incorrect path syntax) (due to the root folder containing almost exclusively symbolic links with invalid targets, I made that faulty assumption). No freeze with `.` as target either. I entered the top-level symbolic link of the root folder, went to multiple depths; traversed multiple symbolic links and folders until I found the troubled symbolic link. I isolated it to a new folder, stripped its path bit by bit, tested each time, until there was no freeze. **Result**: a freeze occurs with a root folder containing at least one symbolic link with "parent ups" `..` and `..\` and `..\..\` as target, which creates a link loop of type _multiple user actions finite loop_. Do I still need to create a new issue?

@pryrt I meant by "which ones do you like?": which definition(s) or type(s) from one definition did you use when you said (paraphrase) that "link loops are not useful"? In other words, which definition's referred meaning or type(s) from one definition's referred meaning do you consider useful? I consider link loop of type _multiple user actions finite loop_ very useful, and the other not useful at all; I am guessing that anyone will agree on having no use for the type _single user action infinite loop_. I am not sure what the people who will work on this issue believe about the one I find useful, and whether they will need some arguments from me for motivation (I assumed so after you said "link loops are not useful" and assumed an implied community wide belief, but now I am not sure what you meant, so I will wait for that; anyway, I have written some arguments for _multiple user actions finite loop_ that I can share if asked by anyone).

Does testing a fix mostly involve downloading a program, running, observing and reporting back?

**Comment by pryrt** [comment]
I said nothing about "useful". I used the word "invalid", but _not_ in regard to the circular linking.  Specifically, I said, 

> Is your problem with the symbolic links only when there is one or more link that points to an invalid target? Or does it also happen when the only symbolic link has a valid target? If it's the latter, then that is definitely a bug. If it's just with an invalid target (or the circular target from earlier in the discussion)

My use of "invalid" target was something like a shortcut or reparse-point linking to a file that doesn't (currently) exist; and that's how I thought the term was used previously in the discussion.  And that could be "useful", even when it's invalid, because someone might want to link to a network resource that's only sometimes there, for example.

My use of "circular" was _separate_ from my use of "invalid" (and, again, I never said "useful", nor anything that could be paraphrased as "useful" IMO).

It took me quite a few reads and thinking about "single user action" and "multiple user actions", because I don't think about links/parse-points in terms of "actions".  I think about what they're actually pointing to.

- When I was thinking "circular", as mentioned previously in the discussion, I was assuming that `a.lnk` points to `b.lnk` and `b.lnk` points to `a.lnk` (where the `*.lnk` are either shorcut files or a shorthand notation for one of the reparse-point implementations), so there's an infinite loop that the OS cannot resolve.  I believe this is what you meant by "single user action".  I technically said nothing abut whether this was "useful" or whether it was "valid", but if you wanted my opinion/interpretation, I don't think it is either. 

- If I have understood your "multiple user actions" correctly, after a lot of thought, the way I would describe it would be: there are two real directories, `c:\path-to\dir.a\` and `c:\another\path\to\dir.b\`.  Inside dir.a, there is a shortcut/reparse-point that links to `dir.b\`, and there is a shortcut/reparse-point inside dir.b which links to `dir.a\`.  I would say this situation is both "useful" and "valid".

And again, this is tangential.  What you need to do (as re-iterated by @donho) is to create a new issue, and give the exact steps needed to replicate the problem, including the commands to create whatever "circular" or "invalid" links are causing the issue with FaW for you.  Just describe all the steps necessary to show your problem.  In the new issue.

**Comment by donho** [comment]
@darthglowball 
> Do I still need to create a new issue?

Yes. A new issue with the detailed info and step by step instructions will help me to reproduce the problem you encountered.
Once I can reproduce the issue, there are more chances to fix the bug.
